File: tools/convert_blog_post_ts/file_object.ts
// import { stack } from "d3";
import type { Block as PbBlock } from "../../protos/anytype/models";
import { SmartBlockType } from "../../protos/anytype/models";

import { SnapshotWithType } from "../../protos/anytype/snapshot";
import { castToAttributeMap, type AttributeMap } from "./attribute";
// import type { Collection, CollectionId } from "./collection";
import { getFieldValue, getShortenId, GLOBAL_RELATION_IDMAP } from "./common";
import type { DataMap } from "./common";
// import type { ContentBlockList, ContentBlockMap } from "./content_block";
// import * as ContentBlock from "./content_block";
import { ObjectTypes } from "./enum_token";
// import { ExternalBookmarkLink } from "./external_link";
// import type { TagList, TagMap, Tag, TagOptionLabel } from "./tag";
// import type { UserId } from "./user";
// import type { WorkspaceId } from "./workspace";
// import { last } from "lodash-es";
//

export interface FileObject {
  id: string;
  _sid: string;
  title: string;
  fileUrl: string;
  fileExt: string;
  fileType: string;
  attributes: DataMap;
}

export const getFileObjects = (
  snapshot_list: SnapshotWithType[],
): FileObject[] => {
  return snapshot_list
    .filter(
      (snapshot) =>
        snapshot.sbType === SmartBlockType.File ||
        snapshot.sbType === SmartBlockType.FileObject,
    )
    .map(fromAnytype);
};

function fromAnytype(raw: SnapshotWithType): FileObject {
  const data = raw.snapshot?.data;
  const detailMap = data?.details;
  const fileExt = getFieldValue<string>(detailMap, "fileExt") ?? "";
  const fileMimeType = getFieldValue<string>(detailMap, "fileMimeType") ?? "";

  const tmp: FileObject = {
    id: getFieldValue(detailMap, "id") ?? "",
    _sid: getShortenId(raw),
    title: getFieldValue(detailMap, "name") ?? "",
    attributes: detailMap ?? {},
    fileUrl: getFieldValue(detailMap, "source") ?? "",
    fileExt,
    fileType: resolveFileType(fileExt, fileMimeType),
  };

  return tmp;
}

const resolveFileType = (fileExt: string, fileMimeType: string): string => {
  if (fileMimeType.startsWith("image")) return "images";
  if (fileMimeType.startsWith("video")) return "videos";
  if (fileMimeType.startsWith("audio")) return "audios";
  // if (fileMimeType.startsWith('document')) return "Document";
  if (fileExt === "pdf") return "pdf";
  if (
    fileExt === "doc" ||
    fileExt === "docx" ||
    fileExt === "xls" ||
    fileExt === "xlsx" ||
    fileExt === "ppt" ||
    fileExt === "pptx"
  )
    return "documents";

  if (fileExt === "txt") return "text";

  if (fileExt === "csv" || fileExt === "json" || fileExt === "xml")
    return "data";

  if (fileExt === "ipynb") return "notebook";

  return "raw";
};

export const exportForMetaData = (fileList: FileObject[]) => {
  const images = fileList
    .filter((f) => f.fileType === "images")
    .map(
      (e) =>
        ({
          url: `/blog/assets/${e.fileUrl}`,
          alt: e.title,
          height: e.attributes["heightInPixels"],
          width: e.attributes["widthInPixels"],
          type: e.attributes["fileMimeType"],
        }) as OpenGraphImage,
    );

  const videos = fileList
    .filter((f) => f.fileType === "videos")
    .map(
      (e) =>
        ({
          url: `/blog/assets/${e.fileUrl}`,
          type: e.attributes["fileMimeType"],
          height: e.attributes["heightInPixels"],
          width: e.attributes["widthInPixels"],
        }) as OpenGraphVideos,
    );
  const audios = fileList
    .filter((f) => f.fileType === "audios")
    .map(
      (e) =>
        ({
          url: `/blog/assets/${e.fileUrl}`,
          type: e.attributes["fileMimeType"],
        }) as OpenGraphAudio,
    );

  let result = {};
  if (images.length > 0) {
    // result.images = images;
    result = Object.assign(result, { images });
  }
  if (videos.length > 0) {
    result = Object.assign(result, { videos });
  }
  if (audios.length > 0) {
    result = Object.assign(result, { audios });
  }

  return result;
};

export interface OpenGraphImage {
  url: string;
  secureUrl?: string;
  type?: string;
  width?: number;
  height?: number;
  alt?: string;
}

export interface OpenGraphVideos {
  url: string;
  secureUrl?: string;
  type?: string;
  width?: number;
  height?: number;
}

export interface OpenGraphAudio {
  url: string;
  secureUrl?: string;
  type?: string;
}

File: tools/convert_blog_post/src/export_model/file_object.rs
use convert_blog_post_marco::set_field_value;
use serde::{Deserialize, Serialize};

use crate::proto::anytype::SnapshotWithType;
use anyhow::anyhow;

use super::{
    common::{AttributeMap, get_field_value, get_shorten_id, get_snapshot_shorthanded},
    page::meta::PageMetaOpenGraphObj,
    trait_impl::{FromRaw, FromSnapshotList},
};

#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FileObject {
    pub id: String,
    #[serde(rename = "_sid")]
    pub sid: String,
    pub title: String,
    pub file_url: String,
    pub file_ext: String,
    pub file_type: String,
    pub file_mime: String,
    pub width: i64,
    pub height: i64,
    pub attributes: AttributeMap,
}

impl<'a> FromRaw<SnapshotWithType<'a>> for FileObject {
    fn from_raw(raw_obj: &SnapshotWithType<'a>) -> Result<FileObject, anyhow::Error> {
        let mut tmp = Self::default();
        let instance = get_snapshot_shorthanded(raw_obj);
        if let Err(e) = instance {
            return Err(anyhow!("fail field map : {:?}", e));
        }
        let (_, field_map) = instance.unwrap();

        let mut id = String::new();
        set_field_value!(id, field_map, "id");
        tmp.id = id.to_owned();
        tmp.sid = get_shorten_id(&id);
        set_field_value!(tmp.title, field_map, "name");

        set_field_value!(tmp.file_url, field_map, "source");
        let mut file_mime_type = String::new();
        set_field_value!(file_mime_type, field_map, "fileMimeType");

        let mut file_ext = String::new();
        set_field_value!(file_ext, field_map, "fileExt");

        let mut width: f64 = 0.0;
        set_field_value!(width, field_map, "widthInPixels");
        tmp.width = width as i64;

        let mut height: f64 = 0.0;
        set_field_value!(height, field_map, "heightInPixels");
        tmp.height = height as i64;

        // println!("field_map : {:#?}", field_map);

        tmp.file_type = resolve_file_type(&file_ext, &file_mime_type).to_owned();
        tmp.file_mime = file_mime_type;
        tmp.file_ext = file_ext;
        tmp.attributes = field_map;
        return Ok(tmp);
    }
}

impl FromSnapshotList for FileObject {
    fn from_snapshot_list(
        list_raw: Vec<crate::proto::anytype::SnapshotWithType>,
    ) -> Result<Vec<FileObject>, anyhow::Error> {
        let mut tmp_list: Vec<FileObject> = vec![];
        for item in list_raw {
            let tmp = FileObject::from_raw(&item);
            if let Err(e) = tmp {
                return Err(anyhow!(e));
            }
            tmp_list.push(tmp.unwrap());
        }

        return Ok(tmp_list);
    }
}

fn resolve_file_type(file_ext: &str, file_mime_type: &str) -> &'static str {
    if file_mime_type.starts_with("image") {
        return "images";
    }
    if file_mime_type.starts_with("video") {
        return "videos";
    }
    if file_mime_type.starts_with("audio") {
        return "audios";
    }
    // if (fileMimeType.startsWith('document')) return "Document";
    if file_ext == "pdf" {
        return "pdf";
    }
    if file_ext == "doc"
        || file_ext == "docx"
        || file_ext == "xls"
        || file_ext == "xlsx"
        || file_ext == "ppt"
        || file_ext == "pptx"
    {
        return "documents";
    }

    if file_ext == "txt" {
        return "text";
    }

    if file_ext == "csv" || file_ext == "json" || file_ext == "xml" {
        return "data";
    }

    if file_ext == "ipynb" {
        return "notebook";
    }

    return "raw";
}

impl FileObject {
    pub fn to_page_meta_og(&self) -> PageMetaOpenGraphObj {
        let mut tmp = PageMetaOpenGraphObj {
            url: self.file_url.to_string(),
            // secure_url: "",
            file_type: Some(self.file_mime.to_string()),
            // width: Some(self.width),
            // height: Some(self.height),
            // alt: Some(self.title.to_string()),
            ..PageMetaOpenGraphObj::default()
        };

        if self.file_type != "audio" {
            tmp.width = Some(self.width);
            tmp.height = Some(self.height);
            tmp.alt = Some(self.title.to_string());
        }

        return tmp;
    }
}
